File: 20-valid-parentheses/valid-parentheses.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        if ((n & 1) == 1) return false; // odd length can't be balanced

        Deque<Character> stack = new ArrayDeque<>(n);
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            switch (c) {
                case '(': case '{': case '[':
                    stack.push(c);
                    break;
                case ')':
                    if (stack.isEmpty() || stack.pop() != '(') return false;
                    break;
                case '}':
                    if (stack.isEmpty() || stack.pop() != '{') return false;
                    break;
                case ']':
                    if (stack.isEmpty() || stack.pop() != '[') return false;
                    break;
                default:
                    // If inputs are guaranteed to be only (){}[], you can omit this.
                    return false;
            }
        }
        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        if len(s)%2 != 0: 
            return False

        def paramatch(b) -> bool:
            if len(stack) == 0:
                return False
            if b == ')' and stack[len(stack)-1] == '(':
                return True
            if b == '}' and stack[len(stack)-1] == '{':
                return True
            if b == ']' and stack[len(stack)-1] == '[':
                return True
            return False
        
        for b in s:
            if b in ['(', '{', '[']:
                stack.append(b)
            elif paramatch(b):
                stack.pop(len(stack)-1)
            else:
                return False

        return len(stack) == 0